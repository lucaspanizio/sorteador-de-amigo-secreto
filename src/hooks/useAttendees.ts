import { useRecoilState } from 'recoil';
import { attendeesState, errorState, resultState } from '@/store/drawer';
import { useError } from './useError';

export const useAttendees = () => {
  const [attendees, setAttendees] = useRecoilState(attendeesState);
  const { updateError } = useError();

  const addAttendee = (name: string) => {
    if (attendees.includes(name)) {
      updateError('Nomes duplicados nÃ£o sÃ£o permitidos!');
      setTimeout(() => updateError(''), 2000);
      return;
    }
    setAttendees((attendees) => [...attendees, name]);
  };

  return { attendees, addAttendee };
};
