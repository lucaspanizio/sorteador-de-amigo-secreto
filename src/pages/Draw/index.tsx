export const Draw = () => {
  return <>Sorteio</>;
};
