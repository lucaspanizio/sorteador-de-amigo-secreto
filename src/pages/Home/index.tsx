import { Header } from '@/components/Header';
import { Form } from '@/components/Form';
import { AttendeesList } from '@/components/AttendeesList';

export const Home: React.FC = () => {
  return (
    <>
      <Header />
      <Form />
      <AttendeesList />
    </>
  );
};
