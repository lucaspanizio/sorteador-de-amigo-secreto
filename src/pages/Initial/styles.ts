import styled from 'styled-components';

export const Button = styled.button`
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 2rem;
  width: 300px;
  height: 80px;
  font-size: 20px;
  box-shadow: 2px 2px 0px 1px #000000;
  border-radius: 45px;
  background-color: #fe652b;
  color: #fff;
  cursor: pointer;

  padding-left: 60px;
  background-image: url('/images/play.png');
  background-size: 40px 40px;
  background-repeat: no-repeat;
  background-position: 24px center;

  &:disabled {
    opacity: 0.6;
    cursor: not-allowed;
  }

  &:hover {
    background-color: #4b69fd;
  }

  @media screen and (max-width: 800px) {
    width: 220px;
    margin: 32px 0;
  }
`;

export const Footer = styled.footer`
  display: flex;
  align-items: center;
  justify-content: space-between;
`;
