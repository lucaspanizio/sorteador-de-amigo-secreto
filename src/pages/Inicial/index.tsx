import { Form } from '@/components/Form';
import { AttendeesList } from '@/components/AttendeesList';
import { Button } from '@/components/Button';
import * as S from './styles';

export const Inicial: React.FC = () => {
  return (
    <section>
      <h2>Vamos come√ßar!</h2>
      <Form />
      <AttendeesList />
      <S.Footer>
        <Button />
        <img src="/images/sacolas.png" alt="Sacolas de compras" />
      </S.Footer>
    </section>
  );
};
