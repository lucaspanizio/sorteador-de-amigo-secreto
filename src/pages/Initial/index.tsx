import { Form } from '@/components/Form';
import { AttendeesList } from '@/components/AttendeesList';
import { useAttendees } from '@/hooks/useAttendees';
import { useNavigate } from 'react-router-dom';
import { useDrawer } from '@/hooks/useDrawer';
import * as S from './styles';

export const Initial = () => {
  const { attendees } = useAttendees();
  const navigate = useNavigate();
  const draw = useDrawer();

  const startGame = () => {
    draw();
    navigate('/sorteio');
  };

  return (
    <section>
      <h1>Vamos come√ßar!</h1>
      <Form />
      <AttendeesList />
      <S.Footer>
        <S.Button
          data-testid="btnIniciar"
          disabled={attendees.length < 3}
          onClick={startGame}
        >
          Iniciar brincadeira!
        </S.Button>
        <img src="/images/sacolas.png" alt="Sacolas de compras" width="150px" />
      </S.Footer>
    </section>
  );
};
