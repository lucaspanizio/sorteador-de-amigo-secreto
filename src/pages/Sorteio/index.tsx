import { FormEvent, useState } from 'react';
import { useAttendees } from '@/hooks/useAttendees';
import { useResult } from '@/hooks/useResult';
import * as S from './styles';

export const Sorteio = () => {
  const { result } = useResult();
  const { attendees } = useAttendees();

  const [actualAttendee, setActualAttendee] = useState('');
  const [secretSanta, setSecretSanta] = useState('');

  const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    if (result?.has(actualAttendee)) {
      setSecretSanta(result.get(actualAttendee)!);

      setTimeout(() => {
        setSecretSanta('');
        setActualAttendee('');
      }, 5000);
    }
  };

  return (
    <section>
      <h1>Quem vai tirar o papelzinho?</h1>
      <form onSubmit={handleSubmit}>
        <S.Select
          value={actualAttendee}
          onChange={({ target }) => setActualAttendee(target.value)}
        >
          <option value="">Selecione um nome</option>
          {attendees.map((name) => (
            <option key={name} value={name}>
              {name}
            </option>
          ))}
        </S.Select>

        <S.OrientationText>
          Clique em sortear para ver quem será seu amigo secreto!
        </S.OrientationText>

        {actualAttendee && secretSanta && (
          <S.SecretSanta role="alert">{secretSanta}</S.SecretSanta>
        )}

        <S.Footer>
          <S.Button type="submit" disabled={!actualAttendee}>
            <img
              src="/images/dado-branco.png"
              alt="Desenho de um dado branco"
              width="30px"
            />
            <span>Sortear!</span>
          </S.Button>
          <img src="/images/aviao.png" alt="Desenho de um avião de papel" />
        </S.Footer>
      </form>
    </section>
  );
};
