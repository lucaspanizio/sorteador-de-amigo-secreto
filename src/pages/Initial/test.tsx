import { RecoilRoot } from 'recoil';
import { fireEvent, render, screen } from '@testing-library/react';
import { useAttendees } from '@/hooks/useAttendees';
import { Initial } from '@/pages/Initial';

const mockNavigation = jest.fn();
const mockDrawer = jest.fn();

jest.mock('react-router-dom', () => {
  return {
    useNavigate: () => mockNavigation,
  };
});

jest.mock('@/hooks/useDrawer', () => {
  return {
    useDrawer: () => mockDrawer,
  };
});

jest.mock('@/hooks/useAttendees', () => {
  return {
    useAttendees: jest.fn(),
  };
});

const renderComponent = () => {
  return render(
    <RecoilRoot>
      <Initial />
    </RecoilRoot>,
  );
};

describe('Page: Initial', () => {
  describe('When render page', () => {
    it('should be exactly like the snapshot', () => {
      (useAttendees as jest.Mock).mockReturnValue({ attendees: [] });
      const { container } = renderComponent();
      expect(container).toMatchSnapshot();
    });
  });

  describe('When there are not enough attendees', () => {
    it('the game can NOT be started', () => {
      (useAttendees as jest.Mock).mockReturnValue({ attendees: [] });

      renderComponent();

      const button = screen.getByTestId('btnIniciar');
      expect(button).toBeDisabled();
    });
  });

  describe('When there are enough attendees', () => {
    beforeEach(() => {
      (useAttendees as jest.Mock).mockReturnValue({
        attendees: ['Ana', 'JoÃ£o', 'Pedro'],
      });

      renderComponent();
    });

    it('the game CAN be started', () => {
      const button = screen.getByTestId('btnIniciar');
      expect(button).not.toBeDisabled();
    });

    it('the game WAS started', () => {
      const button = screen.getByTestId('btnIniciar');
      fireEvent.click(button);

      expect(mockNavigation).toHaveBeenCalledTimes(1);
      expect(mockNavigation).toHaveBeenCalledWith('/sorteio');
      expect(mockDrawer).toHaveBeenCalledTimes(1);
    });
  });
});
