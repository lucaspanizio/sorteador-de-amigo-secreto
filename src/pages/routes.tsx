import { createBrowserRouter, Navigate } from 'react-router-dom';
import { Layout } from '@/components/Layout';
import { Initial } from '@/pages/Initial';
import { Sorteio } from '@/pages/Sorteio';

export const router = createBrowserRouter([
  {
    path: '/',
    element: <Layout />,
    children: [
      { path: '', element: <Initial /> },
      { path: 'sorteio', element: <Sorteio /> },
    ],
  },
]);
