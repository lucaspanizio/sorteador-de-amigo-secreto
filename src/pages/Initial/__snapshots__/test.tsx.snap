// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page: Initial When render page should be exactly like the snapshot 1`] = `
<div>
  <section>
    <h1>
      Vamos come√ßar!
    </h1>
    <form
      class="sc-blHHSb opeqX"
    >
      <input
        class="sc-gtLWhw ggfSGj"
        placeholder="Insira os nomes dos participantes"
        type="text"
        value=""
      />
      <button
        class="sc-egkSDF iDlEkV"
        disabled=""
        type="submit"
      >
        Adicionar
      </button>
    </form>
    <footer
      class="sc-ixGGxD gZDtne"
    >
      <button
        class="sc-kLhKbu fGqRoi"
        data-testid="btnIniciar"
        disabled=""
      >
        Iniciar brincadeira!
      </button>
      <img
        alt="Sacolas de compras"
        src="/images/sacolas.png"
        width="150px"
      />
    </footer>
  </section>
</div>
`;
